File: 145.二叉树的后序遍历.swift
/*
 * @lc app=leetcode.cn id=145 lang=swift
 *
 * [145] 二叉树的后序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
class Solution {
    func postorderTraversal(_ root: TreeNode?) -> [Int] {
        var result = [Int]()
        postorderTraversal(root, result: &result)
        return result
    }
    
    func postorderTraversal(_ node: TreeNode?, result: inout [Int]){
        guard let node = node else{
            return
        }
        
        postorderTraversal(node.left, result: &result)
        postorderTraversal(node.right, result: &result)
        result.append(node.val)
    }
}
// @lc code=end

File: 145.二叉树的后序遍历.py
#
# @lc app=leetcode.cn id=145 lang=python3
#
# [145] 二叉树的后序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        result = []
        if root is None:
            return result
        
        stack = []
        current = root
        # 后续遍历，要等右节点访问后才能访问根节点，用lastVisited来区分右节点是否已经访问过了
        lastVisited = None   
        while current or stack:
            while current:
                stack.append(current)
                current = current.left
        
            top = stack[-1]
            # 没有右子树，或者右子树已经访问过了
            if top.right is None or top.right == lastVisited:
                result.append(top.val)
                lastVisited = top
                stack.pop()
                current = None
            else:
                current = top.right
            
        return result
    
# @lc code=end

